### DISCUSSION TITLE
`looks like we got no XML document` - nice message, but why is it so? 🤔

### DISCUSSION THREAD

**Discussion by trailsnail** [phpro/soap-client#505] [Q&A] [answered]
# What is it about?
We retrieve data via a SOAP api, a list of many offers.
These offers are created and maintained in the customer's ERP system. Employees sometimes copy text from Word or a PDF into the text fields of a offer. Sometimes special characters are inserted.

![grafik](https://github.com/phpro/soap-client/assets/4742603/c2e6525a-6505-473c-8b7e-4872bff9c6fe)

When decoding the request response in the `ExtSoapDecoder::decode()`, the error `looks like we got no XML document` occurs. The entire process aborts. The error-free offers cannot be processed due to an incorrect offer.

Unfortunately, it is not possible to retrieve each offer individually in order to bypass the incorrect one.

# What can we do?
Debugging is currently the order of the day, narrowing down the period from when the error occurs and which offers could match it. Then retrieve the offer explicitly via Postman or debugging in the code, copy the response for decoding into the IDE or examine it with text editors to find the needle in the haystack that makes decoding impossible.

# Is there another way?
It would be helpful to have an error output that provides us with information about where the error is in the data or even what exactly is causing the error.
Do you also struggle with such challenges or have you already found a solution? We would be delighted to hear your ideas and suggestions.

Kind regards

**Comment by veewee** [comment, marked as answer]
Hello,

The exception is coming from PHP's SoapClient and means it is unable to parse the provided XML response.

`&#x2;` is an invalid character in XML. The best way to overcome this error is to add a HTTP plugin/Middleware that makes sure the XML is parsable by for example cleaning up this and any other unsupported character you encounter before they get decoded by the SoapClient.

https://3v4l.org/PkH7j

```php
$doc = new DOMDocument();
$doc->loadXML('<test>wor&#x2;ld</test>');

echo($doc->documentElement->nodeValue);
```

> Warning: DOMDocument::loadXML(): xmlParseCharRef: invalid xmlChar value 2 in Entity,

**Reply by trailsnail** [reply]
Thank you. We will have a look into this 👀

**Reply by trailsnail** [reply]
Thanks again for the tip ❤️ 

We now parse the response before Client decoding with DOMDocument via middleware.
If an error occurs, various regexes are used to determine details from the XML string.
Starting from the error position, we take a substring and determine the XML entities with their content in order to enrich the notification with helpful details.

The `XMLReader` was unfortunately not an option, as it already removes invalid entities from the response during initialization - I was surprised that the XML was generally valid when I tried it. 

```php

// Middleware
public function afterResponse(ResponseInterface $response): ResponseInterface
{
    try {
        $this->parseXmlPreflight($response->getBody()->getContents());
    } catch (Throwable $e) {
        // do nothing atm
    }

    return parent::afterResponse($response);
}

private function parseXmlPreflight(string $xmlString): void
{
	$dom = new DOMDocument();
	libxml_use_internal_errors(true); // Enable error handling
	$dom->loadXML($xmlString, LIBXML_NOERROR);

	$errors = libxml_get_errors();
	if ($errors === []) {
		return;
	}

	$message = new MattermostMessage('Fehler bei der Verarbeitung von XML-Daten', ':rotating_light:', true);

	$this->enrichUserRequestDetails($message);

	foreach($errors as $error) {
		$this->collectErrorDetails($message, $error, $xmlString); // Try to extract details from XML to enrich the notification
	}

	$message->notifyPhilippTeamInternal();
}

```

![grafik](https://github.com/phpro/soap-client/assets/4742603/53684d03-117b-4fae-9036-6ab107865a8e)

**Reply by veewee** [reply]
That's some sweet error handling! 😮 Nice.

**Reply by trailsnail** [reply]
@veewee 

Time has come, project is moving to PHP `8.3`. We got Soap client running in version `2.0` (migration not ready).
The original answer to the "check before decode" was to use a `afterResponse()` Hook in our custom middleware.

Now in new version, the engine is generated like following:

```php
return DefaultEngineFactory::create(
            ExtSoapOptions::defaults($wsdl, ['soap_version' => SOAP_1_2])
                ->withClassMap($classMapCollection)
                ->withWsdlProvider(
                    new TemporaryWsdlLoaderProvider(
                        Psr18Loader::createForClient(
                            $httpClient,
                        ),
                        new Md5Strategy(),
                        sys_get_temp_dir().DIRECTORY_SEPARATOR.SOAP_TMP_PATH
                    )
                ),
            Psr18Transport::createForClient(
                new PluginClient(
                    $httpClient,
                    [
                        new WebconnectorHostRewritePlugin(),
                        new WebconnectorRequestStopWatch(),
                        new WsaMiddleware2005(WsaMiddleware2005::WSA_ADDRESS2005_ANONYMOUS)
                    ]
                )
            ),
        );

```

How is it possible now to write a Plugin, which can handle the response before it is decoded?

Maybe you have an idea 😉 

Kind regards

**Reply by veewee** [reply]
Hello,

The middleware system of v1 got removed in favour for regular php-http plugins.

More info:
- https://docs.php-http.org/en/latest/plugins/build-your-own.html
- https://github.com/php-soap/psr18-transport?tab=readme-ov-file#middleware

You should be able to transform the Middleware into a Plugin and add it to the PluginClient:

```php
            Psr18Transport::createForClient(
                new PluginClient(
                    $httpClient,
                    [
                        new WebconnectorHostRewritePlugin(),
                        new WebconnectorRequestStopWatch(),
                        new WsaMiddleware2005(WsaMiddleware2005::WSA_ADDRESS2005_ANONYMOUS),

                        // > Here:
                        new YourCustomCheckPlugin(),


                    ]
                )
            ),
```

**Reply by trailsnail** [reply]
Thanks, I was on this way already and tried my own plugin as last in chain.
I thought, I found the solution in [demo code here](https://docs.php-http.org/en/latest/plugins/build-your-own.html), wher I would have the `$response` to do something with.

```php
public function handleRequest(RequestInterface $request, callable $next, callable $first): Promise
    {
        return $next($request)->then(function (ResponseInterface $response) {
            try {
                // Check response for possible invalid characters 
            } catch (Throwable $e) {
                // do nothing atm
            }

            return $response;
        }, function (\Http\Client\Exception $exception) {
            echo $exception->getMessage();

            throw $exception;
        });
    }

```

But already the call of `return $next($request)->then( ...` results in exactly the error, this discussion was for 🤣  

```
Uncaught exception: 'Phpro\SoapClient\Exception\SoapException' with message 'looks like we got no XML document' (0)
> in /cluster/www/www/www/philipp/vendor/phpro/soap-client/src/Phpro/SoapClient/Exception/SoapException.php:19
Caused by 'SoapFault' with message 'looks like we got no XML document' (0)
> in /cluster/www/www/www/philipp/vendor/php-soap/ext-soap-engine/src/ExtSoapDecoder.php:26
Stack trace:
#0 /cluster/www/www/www/philipp/vendor/phpro/soap-client/src/Phpro/SoapClient/Caller/EngineCaller.php(37): Phpro\SoapClient\Exception\SoapException::fromThrowable()
#1 /cluster/www/www/www/philipp/vendor/phpro/soap-client/src/Phpro/SoapClient/Caller/EventDispatchingCaller.php(30): Phpro\SoapClient\Caller\EngineCaller->__invoke()

```

So any other ideas? 🙏🏻

**Reply by veewee** [reply]
Since the error is only thrown in the `ExtSoapDecoder` it means that the plugins are already executed at that point.
Your plugin probably gets bypassed by another one?

**Reply by trailsnail** [reply]
No, the only place where I pass the plugin is when creating the engine.

### CustomerClientFactory
```php
class CustomerClientFactory
{
    public static function factory(string $wsdl) : CustomerClient
    {
        $engine = SoapConfig::getEngine($wsdl, CustomerClassmap::getCollection());  

        $eventDispatcher = new EventDispatcher();
        $caller = new EventDispatchingCaller(new EngineCaller($engine), $eventDispatcher);

        return new CustomerClient($caller);
    }
}

```

### SoapConfig for getting the Engine
```php

public static function getEngine(string $wsdl, ClassMapCollection $classMapCollection): Engine
    {
        $httpClient = Psr18ClientDiscovery::find();
        return DefaultEngineFactory::create(
            ExtSoapOptions::defaults($wsdl, ['soap_version' => SOAP_1_2])
                ->withClassMap($classMapCollection)
                ->withWsdlProvider(
                    new TemporaryWsdlLoaderProvider(
                        Psr18Loader::createForClient(
                            $httpClient,
                        ),
                        new Md5Strategy(),
                        sys_get_temp_dir().DIRECTORY_SEPARATOR.SOAP_TMP_PATH
                    )
                ),
            Psr18Transport::createForClient(
                new PluginClient(
                    $httpClient,
                    [
                        new WebconnectorHostRewritePlugin(),
                        new WebconnectorRequestStopWatch(),
                        new WsaMiddleware2005(WsaMiddleware2005::WSA_ADDRESS2005_ANONYMOUS),
                        new \WebconnectorXmlErrorDecoder() // <= this is the class which should handle the response before it is decoded
                    ]
                )
            ),
        );
    }

```

### WebconnectorXmlErrorDecoder
```php

final class WebconnectorXmlErrorDecoder implements Plugin
{
    public function handleRequest(RequestInterface $request, callable $next, callable $first): Promise
    {
        return $next($request)->then(function (ResponseInterface $response) {

            try {
                // Check response for possible invalid characters before it is decoded - but ExtSoapDecoder already touched it
            } catch (Throwable $e) {
                // do nothing atm
            }

            return $response;
        }, function (\Http\Client\Exception $exception) {
            echo $exception->getMessage();

            throw $exception;
        });
    }
}
```

I tried to do something like in class ` Soap\Psr18WsseMiddleware\WsseMiddleware`, adding a `afterResponse` hook as it was working in the middleware system of v1, but the `afterResponse` is now called after the broken Promise `$next($request)->then`.

All examples shown in https://docs.php-http.org/en/latest/plugins/build-your-own.html are made for successful decodable responses, but not for those who contains invalid characters 🤣  - that is what I think atm.

**Reply by veewee** [reply]
Isn't the issue related to:

```php
            } catch (Throwable $e) {
                // do nothing atm
            }
```

This results in the `response` being passed as-is with invalid XML chars.
You should be able to throw an exception in there which will result in the response not being passed to the SOAP decoder.


(note that you are not required to include an error callback in `then(_, here)`)

**Reply by veewee** [reply]
BTW, moving from the old middleware to plugin, should be a matter of using this method - which is basically what the base Middleware class was providing : https://github.com/phpro/soap-client/blob/v1.7.x/src/Phpro/SoapClient/Middleware/Middleware.php

```php
    public function handleRequest(RequestInterface $request, callable $next, callable $first): Promise
    {
        return $this->beforeRequest($next, $request)
            ->then(
                (function (ResponseInterface $response) {
                    return $this->afterResponse($response);
                })->bindTo($this),
                (function (Exception $exception) {
                    $this->onError($exception);
                })->bindTo($this)
            );
    }
```

(since `interface MiddlewareInterface extends Plugin`)

**Reply by trailsnail** [reply]
> This results in the `response` being passed as-is with invalid XML chars.
> You should be able to throw an exception in there which will result in the response not being passed to the SOAP decoder.

No, this doesn't work.

Here is a little bit more stacktrace from error:

```
Uncaught exception: 'Phpro\SoapClient\Exception\SoapException' with message 'looks like we got no XML document' (0)
> in /cluster/www/www/www/philipp/vendor/phpro/soap-client/src/Phpro/SoapClient/Exception/SoapException.php:19
Caused by 'SoapFault' with message 'looks like we got no XML document' (0)
> in /cluster/www/www/www/philipp/vendor/php-soap/ext-soap-engine/src/ExtSoapDecoder.php:26
Stack trace:
#0 /cluster/www/www/www/philipp/vendor/phpro/soap-client/src/Phpro/SoapClient/Caller/EngineCaller.php(37): Phpro\SoapClient\Exception\SoapException::fromThrowable()
#1 /cluster/www/www/www/philipp/vendor/phpro/soap-client/src/Phpro/SoapClient/Caller/EventDispatchingCaller.php(30): Phpro\SoapClient\Caller\EngineCaller->__invoke()
#2 /cluster/www/www/www/philipp/soap/offer/generated/OfferClient.php(60): Phpro\SoapClient\Caller\EventDispatchingCaller->__invoke()
#3 /cluster/www/www/www/philipp/app/portal/lib/navision/integration/OfferRepository.php(244): WebConnector\Offer\OfferClient->readMultiple()
```

### 👀 on working request
```php
 public function handleRequest(RequestInterface $request, callable $next, callable $first): Promise
    {
        return $next($request)->then(function (ResponseInterface $response) {

            dump($response);
            
            try {
                // Check response for possible invalid characters
            } catch (Throwable $e) {
                // do nothing atm
            }

            return $response;
        }, function (\Http\Client\Exception $exception) {
            echo $exception->getMessage();

            throw $exception;
        });
    }
```

Here I have the Response to work with (screenshot dumped `$response`)

![grafik](https://github.com/phpro/soap-client/assets/4742603/47ca5a67-6155-4ff3-be22-2055cb8f40fa)

But the request which receives the XML with invalid character errors so hard that I can't even catch the exception in the middleware.
I don't know how it is possible to do something with the `$response` it is decoded there:

```php
final class SimpleEngine implements Engine
{
	...
    public function request(string $method, array $arguments)
    {
        $request = $this->driver->encode($method, $arguments);
        $response = $this->transport->request($request);

        return $this->driver->decode($method, $response);
    }
	...
}
```

**Reply by veewee** [reply]
Hello,

It's a bit hard for me to see what is going wrong in your very specific code of course :)

Does the invalid request got dumped as well or not?
Given that you state "errors so hard that I can't even catch the exception", it hints me again that the response is not being parsed by this specific middleware. Would it be possible to comment out all other middlewares to make sure none of them are trying to parse the XML and to return the response as-is without skipping any next middleware?

You should be able to dump the exception from the middleware. If not, it's most likely another middleware that is interfering.

**Reply by trailsnail** [reply]
A huge apology from me and sorry for the inconvenience.

The original middleware did not have the purpose of intercepting the error "looks like we got no XML document" and creating corresponding feedback for the user.

Therefore, it is perfectly fine for the error to occur.
What did not work, I believed, was the test decoding to identify possible invalid characters.

But that also worked. What really didn't work was collecting the XML errors via `libxml_use_internal_errors(true)` in our Helper class called witzhin the `afterResponse()` middleware method. The flag `libxml_use_internal_errors` was reset too early, which resulted in `libxml_get_errors()` not returning any errors.

Conclusion: misinterpreted what can be misinterpreted, looked for in the wrong place.

Own middleware with `handleRequest()` and `afterResponse()` hook now works as expected.

Thanks for your support and have a nice time

![grafik](https://github.com/phpro/soap-client/assets/4742603/d8a6fa7c-51dc-4f74-83b4-54740f1d5f67)
